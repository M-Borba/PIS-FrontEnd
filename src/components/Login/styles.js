import { makeStyles } from "@material-ui/core/styles";

export const useStyles = makeStyles((theme) => ({
  paper: {
    margin: theme.spacing(8, 4),
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
  },
  form: {
    width: "100%", // Fix IE 11 issue.
    marginTop: theme.spacing(1),
  },
  submit: {
    marginTop: 20,
  },
  errorMsg: { position: "fixed", color: "red" },
  root: {
    width: "100% !important",
  },
}));
