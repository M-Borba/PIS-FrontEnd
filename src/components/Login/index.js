import React from "react";
import PropTypes from "prop-types";
import Button from "@material-ui/core/Button";
import TextField from "@mui/material/TextField";
import Typography from "@mui/material/Typography";

import { useStyles } from "./styles";
import { LOGIN_LABELS } from "../../config/globalVariables";

Login.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  onInputChange: PropTypes.func,
  email: PropTypes.string.isRequired,
  password: PropTypes.string.isRequired,
  error: PropTypes.string,
};

export default function Login({
  onSubmit,
  onInputChange,
  email,
  password,
  error,
}) {
  const classes = useStyles();

  return (
    <div className={classes.paper} data-testid="login">
      <Typography component="h1" variant="h5">
        {LOGIN_LABELS.INICIAR_SESION}
      </Typography>
      <form className={classes.form} noValidate onSubmit={onSubmit}>
        <TextField
          variant="outlined"
          margin="normal"
          required
          fullWidth
          id="email"
          type="email"
          label={LOGIN_LABELS.EMAIL}
          name="email"
          autoComplete="email"
          value={email}
          onChange={(e) => onInputChange(e)}
          autoFocus
        />
        <TextField
          variant="outlined"
          margin="normal"
          required
          fullWidth
          name="password"
          label={LOGIN_LABELS.CONTRASENA}
          type="password"
          id="password"
          value={password}
          onChange={(e) => onInputChange(e)}
          autoComplete="current-password"
        />
        <Button
          style={{
            marginTop: 20,
          }}
          role="submit"
          type="submit"
          fullWidth
          variant="contained"
        >
          {LOGIN_LABELS.INICIAR_SESION}
        </Button>
        <Typography className={classes.errorMsg} component="h2">
          {error}
        </Typography>
      </form>
    </div>
  );
}
