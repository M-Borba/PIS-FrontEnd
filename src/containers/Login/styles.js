import { makeStyles } from "@material-ui/core/styles";

export const useStyles = makeStyles((theme) => ({
  container: {
    display: "flex",
    height: "100vh",
    justifyContent: "center",
    alignItems: "center",
  },
  imgcontainer: {
    width: theme.typography.pxToRem(200),
    marginTop: theme.typography.pxToRem(40),
  },
  paper: {
    borderRadius: "16px !important",
  },
}));
