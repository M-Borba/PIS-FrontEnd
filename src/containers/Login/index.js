/**
 * Login
 *
 * This container uses the backend api
 * to autenticate.
 *
 */

import React, { useEffect, useState } from "react";
import Paper from "@mui/material/Paper";
import Box from "@mui/material/Box";

import { axiosInstance } from "../../config/axios";
import Login from "../../components/Login";
import ChangePassword from "../../components/LoginChangePassword";
import effectus_wallpaper from "../../resources/effectus_wallpaper.png";
import { useHistory } from "react-router-dom";
import { NOT_LOGGED } from "../../config/globalVariables";
import { useStyles } from "./styles";

export default function LoginView() {
  const history = useHistory();
  const [email, setEmail] = useState("");
  const [loginError, setLoginError] = useState("");
  const [password, setPassword] = useState("");
  const [passwordConfirmation, setPasswordConfirmation] = useState("");
  const [needsPasswordReset, setPasswordReset] = useState(false);
  const classes = useStyles();

  useEffect(() => {
    if (
      localStorage.getItem("uid") !== null &&
      localStorage.getItem("uid") !== undefined &&
      localStorage.getItem("uid") !== NOT_LOGGED &&
      !needsPasswordReset
    ) {
      window.location = "/inicio";
    }
  }, []);

  const signInAndSetHeaders = () =>
    axiosInstance
      .post("/users/sign_in", {
        user: {
          email: email,
          password: password,
        },
      })
      .then((response) => {
        const headers = response.headers;
        localStorage.setItem("token", headers["access-token"]);
        localStorage.setItem("client", headers["client"]);
        localStorage.setItem("uid", headers["uid"]);
        history.push("/Inicio");
      });

  const handleLoginSubmit = (e) => {
    e.preventDefault();
    signInAndSetHeaders().catch((error) => {
      if (error.response?.data?.needs_password_reset === true) {
        const headers = error.response.headers;
        localStorage.setItem("token", headers["access-token"]);
        localStorage.setItem("uid", headers.uid);
        localStorage.setItem("client", headers.client);
        setPassword("");
        setPasswordReset(true);
      } else {
        setLoginError(error.response?.data?.error);
      }
    });
  };

  const handlePasswordChangeSubmit = (e) => {
    e.preventDefault();
    axiosInstance
      .put(
        "/users/password",
        {
          password: password,
          password_confirmation: passwordConfirmation,
        },
        {
          headers: {
            "access-token": localStorage.getItem("token"),
            client: localStorage.getItem("client"),
            uid: localStorage.getItem("uid"),
          },
        }
      )
      .then(() => {
        window.location = "/Inicio";
      })
      .catch((error) => {
        console.error(error.response);
        localStorage.clear();
        setLoginError(error.response?.data?.error);
      });
  };
  const checkInput = (e) => {
    setLoginError("");
    if (e.target.name === "email") setEmail(e.target.value);
    if (e.target.name === "password") setPassword(e.target.value);
    if (e.target.name === "passwordConfirmation")
      setPasswordConfirmation(e.target.value);
  };
  return (
    <Box className={classes.container}>
      <Paper
        variant="elevation"
        elevation={3}
        classes={{ root: classes.paper }}
        // className={classes.paper}
      >
        <Box display="flex" flexDirection="column" alignItems="center">
          <img
            className={classes.imgcontainer}
            src={effectus_wallpaper}
            alt="Logo"
          />
          {!needsPasswordReset ? (
            <Login
              onSubmit={(e) => handleLoginSubmit(e)}
              onInputChange={(e) => checkInput(e)}
              email={email}
              password={password}
              error={loginError}
            />
          ) : (
            <ChangePassword
              onSubmit={(e) => handlePasswordChangeSubmit(e)}
              onInputChange={(e) => checkInput(e)}
              password={password}
              passwordConfirmation={passwordConfirmation}
              error={loginError}
            />
          )}
        </Box>
      </Paper>
    </Box>
  );
}
