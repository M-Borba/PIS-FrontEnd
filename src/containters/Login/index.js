/**
 * Login
 *
 * This container uses the backend api
 * to autenticate.
 *
 */

import React, { useState } from "react";
import { axiosInstance, setHeaders } from "../../config/axios";
import Login from "../../components/Login";
import Grid from "@material-ui/core/Grid";
import effectus_wallpaper from "../../resources/effectus_wallpaper.png";

export default function LoginView() {
  const [email, setEmail] = useState("");
  const [password, setPassowrd] = useState("");

  const handleSubmit = (e) => {
    e.preventDefault();
    if (email == "" || password == "") {
      alert("Completar todos los campos para iniciar sesiÃ³n");
    } else {
      axiosInstance
        .post(
          "/users/sign_in",
          {
            user: {
              email: email,
              password: password,
            },
          }
        )
        .then((response) => {
          let headers = response.headers;
          setHeaders(headers["access-token"], headers.client, headers.uid);
          window.location.reload();
        })
        .catch((error) => {
          console.log(error);
        });
    }
  };
  const checkInput = (e) => {
    e.preventDefault();
    if (e.target.name == "email") setEmail(e.target.value);
    if (e.target.name == "password") setPassowrd(e.target.value);
  };
  return (
    <Grid container spacing={2} justifyContent="center" alignItems="center">
      <Grid item md={7} sm={12}>
        <img
          src={effectus_wallpaper}
          className="logo"
          alt="effectus wallpaper"
          width="100%"
        />
      </Grid>
      <Grid item md={5} sm={12}>
        <Login
          onSubmit={(e) => handleSubmit(e)}
          onInputChange={(e) => checkInput(e)}
          email={email}
          password={password}
        />
      </Grid>
    </Grid>
  );
}
