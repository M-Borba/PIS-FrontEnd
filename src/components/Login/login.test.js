import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Login from "./index";
import React from "react";

describe("login component tests", () => {
  it("renders correctly", () => {
    const mockSubmit = jest.fn((e) => e.preventDefault());
    const mockInputChange = jest.fn((e) => e.preventDefault());
    let email = "";
    let password = "";
    const { container } = render(
      <Login
        onSubmit={(e) => mockSubmit(e)}
        onInputChange={(e) => mockInputChange(e)}
        email={email}
        password={password}
        error={"Identidad o contrase침a no v치lida."}
      />
    );
    //Make sure the error is visible
    expect(
      screen.queryByText("Identidad o contrase침a no v치lida.")
    ).toBeVisible();
    // Matches the last snapshot taken
    expect(container).toMatchSnapshot();
    // calls both functions
    fireEvent(screen.getByRole("submit"), new MouseEvent("click"));

    expect(mockSubmit).toHaveBeenCalled();
  });
});
