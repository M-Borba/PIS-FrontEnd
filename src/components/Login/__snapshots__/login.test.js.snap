// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login component tests renders correctly 1`] = `
<div>
  <div
    class="makeStyles-paper-1"
    data-testid="login"
  >
    <h1
      class="MuiTypography-root MuiTypography-h5"
    >
      Iniciar sesión
    </h1>
    <form
      class="makeStyles-form-2"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-focused Mui-focused Mui-required Mui-required"
          data-shrink="true"
          for="email"
          id="email-label"
        >
          Email
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            autocomplete="email"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="email"
            name="email"
            required=""
            type="email"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-7 PrivateNotchedOutline-legendNotched-8"
            >
              <span>
                Email
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
          for="password"
          id="password-label"
        >
          Contraseña
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            autocomplete="current-password"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="password"
            name="password"
            required=""
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-7"
            >
              <span>
                Contraseña
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-3 MuiButton-containedPrimary MuiButton-fullWidth"
        role="submit"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Iniciar sesión
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <h2
        class="MuiTypography-root makeStyles-errorMsg-4 MuiTypography-body1"
      >
        No se pudo iniciar sesión
      </h2>
      <div
        class="MuiBox-root MuiBox-root-9"
      />
    </form>
  </div>
</div>
`;
